heron/packing/src/java/com/twitter/heron/packing/Container.java

```diff
@@ -23,50 +28,94 @@
     this.maxDisk = resource.getDisk();
+    instances = new HashSet<PackingPlan.InstancePlan>();
   }
 
   /**
    * Check whether the container can accommodate a new instance with specific resource requirements
    *
    * @return true if the container has space otherwise return false
    */
-  private boolean hasSpace(long ram, double cpuCores, long disk) {
-    return usedRam + ram <= maxRam
-        && usedCpuCores + cpuCores <= maxCpuCores
-        && usedDisk + disk <= maxDisk;
+  private boolean hasSpace(Resource resource) {
+    Resource usedResources = this.getTotalUsedResources();
+    return usedResources.getRam() + resource.getRam() <= this.maxRam
+        && usedResources.getCpu() + resource.getCpu() <= this.maxCpuCores
+        && usedResources.getDisk() + resource.getDisk() <= this.maxDisk;
   }
 
   /**
-   * Update the resources currently used by the container, when a new instance with specific
-   * resource requirements has been assigned to the container.
+   * Computes the used resources of the container by taking into account the resources
+   * allocated for each instance.
    *
-   * @return true if the instance can be added to the container, false otherwise
+   * @return a Resource object that describes the used cpu, ram and disk in the container.
    */
-  public boolean add(Resource resource) {
-    if (this.hasSpace(resource.getRam(), resource.getCpu(), resource.getDisk())) {
+  private Resource getTotalUsedResources() {
+    long usedRam = 0;
+    double usedCpuCores = 0;
+    long usedDisk = 0;
+    for (PackingPlan.InstancePlan instancePlan : this.instances) {
+      Resource resource = instancePlan.getResource();
       usedRam += resource.getRam();
       usedCpuCores += resource.getCpu();
       usedDisk += resource.getDisk();
+    }
+    return new Resource(usedCpuCores, usedRam, usedDisk);
+  }
+
+  /**
+   * Update the resources currently used by the container, when a new instance with specific
+   * resource requirements has been assigned to the container.
+   *
+   * @return true if the instance can be added to the container, false otherwise
+   */
+  public boolean add(PackingPlan.InstancePlan instancePlan) {
+    if (this.hasSpace(instancePlan.getResource())) {
+      this.instances.add(instancePlan);
       return true;
     } else {
       return false;
     }
   }
+
+  /**
+   * Remove an instance of a particular component from a container and update its
+   * corresponding resources.
+   *
+   * @return the corresponding instance plan if the instance is removed the container.
+   * Return void if an instance is not found
+   */
+  public Optional<PackingPlan.InstancePlan> removeAnyInstanceOfComponent(String component) {
+    Optional<PackingPlan.InstancePlan> instancePlan = getAnyInstanceOfComponent(component);
+    if (instancePlan.isPresent()) {
+      PackingPlan.InstancePlan plan = instancePlan.get();
+      this.instances.remove(plan);
+      return Optional.of(plan);
```

Comment:
Instead of creating a new Optional you could return the same one removed: `instancePlan`.